Add Z-index to layers

gberaudo opened the pull request:
This commit allows setting a Z-index property on the layers and uses
it in the canvas, WEBGL and DOM map renderers for ordering the layers
before rendering.

Default z-index is `0` for managed layers and `1000` for unmanaged ones.
It allows always on bottom, always on top and more complex layer layouts.

gberaudo commented:
Closes:  #4053

TODO:
- [ ] document the new ordering behaviour;
- [ ] document the special 0 and 1000 values;
- [ ] convert the example to a unit test (check the color of the central point:
   OK if it is red; KO if it is blue).

gberaudo commented:
The PR is now ready for review.

elemoine commented on src/ol/layer/layer.js:
```diff
@@ -152,7 +152,7 @@ ol.layer.Layer.prototype.handleSourcePropertyChange_ = function() {
  * will not filter the layer, and it will be rendered on top. This is useful for
  * temporary layers. To remove an unmanaged layer from the map, use
- * `#setMap(null)`.
+ * `#setMap(null)`. Unmanaged layers behave as if their Z-index is 1000.
```
Does this mean a user is able to place some of his managed layers above unmanaged layers?

gberaudo replied:
Indeed, layers with Z-index > 1000 will be rendered above unmanaged layers.

tschaub commented on src/ol/layer/layer.js:
```diff
@@ -172,6 +172,7 @@ ol.layer.Layer.prototype.setMap = function(map) {
         map, ol.render.EventType.PRECOMPOSE, function(evt) {
           var layerState = this.getLayerState();
           layerState.managed = false;
+          layerState.zIndex = 1000;
```
Maybe `Infinity` would be less arbitrary.

elemoine replied:
I've thought about that too. But using a non-infinity value makes things more flexible to the user, as it is still possible to draw layers above unmanaged layers.

Also, with @gberaudo's current patch, calling `setZIndex` on an unmanaged layer has no effect. I think this should at least me documented. But it may also be great to support calling `setZIndex` on an unmanaged layer. An unmanaged layer would then just be a layer that the map doesn't know about (so you can silently add and remove layers if you want). Maybe I'm going to far here. And if people think I am indeed going too far then I'll go with `Infinity` as suggested by Tim.

gberaudo replied:
There are two arbitrary values, `0` and `1000`, which delimitate three main ranges for the user:
- below 0 for layers to place under all others;
- above 1000 for layers to place on top of all others, including unmanaged layers;
- between 0 and 1000 for placing between them.
The proposition of @elemoine to allow setting Z-index on user unmanaged layers fits well with this vision. It provides the maximum flexibility to the user.

elemoine replied:
> The proposition of @elemoine to allow setting Z-index on user unmanaged layers fits well with this vision. It provides the maximum flexibility to the user.

The implementation is not obvious to me though.

gberaudo replied:
What about the following diff from `setMap()`?
```
-          layerState.zIndex = 1000;
+          var zIndex = layerState.zIndex;
+          layerState.zIndex = zIndex !== 0 ? zIndex : 1000;
```

elemoine replied:
I had assumed that `layer.setZIndex(0)` would work as expected :)

tschaub commented:
> The extended ordering behaviour is documented in ol.layer.Base#setZIndex(), ol.layer.Layer#setMap(). What is the appropriate place to write a detailed explanation?

I think `zIndex` would make sense as a layer option (a new `olx.layer.LayerOption` property), in which case it would be documented [in `olx.js`](https://github.com/openlayers/ol3/blob/f5df71ba7057a738f986903339cc8781b60580cd/externs/olx.js#L3010-L3031).

tschaub commented:
This looks nice @gberaudo.  I look forward to trying it out.

elemoine commented:
> I think zIndex would make sense as a layer option (a new olx.layer.LayerOption property), in which case it would be documented in olx.js.

Agree.

gberaudo commented:
I updated the PR using `Infinity` instead of `1000` for unmanaged layer Z-index and updated documentation in olx externs.
If it is now good for you, please merge.

elemoine commented on externs/olx.js:
```diff
@@ -3091,6 +3102,15 @@ olx.layer.LayerOptions.prototype.extent;
+ * @type {number|undefined}
+ * @api
+ */
+olx.layer.LayerOptions.prototype.zIndex;
```
Shouldn't `zIndex` be an option on `olx.layer.BaseOptions`?

elemoine replied:
@gberaudo just made me realize that `zIndex` in on `BaseOptions` already.

elemoine commented:
Thanks for the excellent work @gberaudo!

gberaudo commented:
Thank you for the review @tschaub and @elemoine.
